File: app/src/main/java/dk/lndesign/relay/model/Stream.java
package dk.lndesign.relay.model;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.annotations.SerializedName;

import java.util.Locale;

/**
 * @author Lars Nielsen <lars@lndesign.dk>
 */
public class Stream implements Parcelable {
    @SerializedName("_id")
    private String id;
    private String game;
    private Integer viewers;
    private Preview preview;
    private Channel channel;
    @SerializedName("stream_type")
    private String streamType;

    public Stream(Parcel in) {
        id = in.readString();
        game = in.readString();
        viewers = in.readInt();
        preview = in.readParcelable(Preview.class.getClassLoader());
        channel = in.readParcelable(Channel.class.getClassLoader());
        streamType = in.readString();
    }

    public String getId() {
        return id;
    }

    public String getGame() {
        return game;
    }

    public Integer getViewers() {
        return viewers;
    }

    public Preview getPreview() {
        return preview;
    }

    public Channel getChannel() {
        return channel;
    }

    public String getStreamType() {
        return streamType;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(getId());
        dest.writeString(getGame());
        dest.writeInt(getViewers());
        dest.writeParcelable(getPreview(), flags);
        dest.writeParcelable(getChannel(), flags);
        dest.writeString(getStreamType());
    }

    public static final Creator<Stream> CREATOR = new Creator<Stream>() {
        public Stream createFromParcel(Parcel in) {
            return new Stream(in);
        }

        public Stream[] newArray(int size) {
            return new Stream[size];
        }
    };

    public String toString() {
        return String.format(Locale.getDefault(),
                "Stream { id: %s, game: '%s', viewers: %d, channel: %s }",
                getId(), getGame(), getViewers(), getChannel());
    }

    public static class Preview implements Parcelable {
        private String small;
        private String medium;
        private String large;

        public Preview(Parcel in) {
            small = in.readString();
            medium = in.readString();
            large = in.readString();
        }

        public String getSmall() {
            return small;
        }

        public String getMedium() {
            return medium;
        }

        public String getLarge() {
            return large;
        }

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeString(getSmall());
            dest.writeString(getMedium());
            dest.writeString(getLarge());
        }

        public static final Creator<Preview> CREATOR = new Creator<Preview>() {
            public Preview createFromParcel(Parcel in) {
                return new Preview(in);
            }

            public Preview[] newArray(int size) {
                return new Preview[size];
            }
        };

        public String toString() {
            return String.format(Locale.getDefault(),
                    "Preview { small: %s, medium: %s, large: %s }",
                    getSmall(), getMedium(), getLarge());
        }
    }

    public static class Channel implements Parcelable {
        @SerializedName("_id")
        private String id;
        @SerializedName("display_name")
        private String displayName;
        private String name;
        private String logo;
        @SerializedName("profile_banner")
        private String profileBanner;

        public Channel(Parcel in) {
            id = in.readString();
            displayName = in.readString();
            name = in.readString();
            logo = in.readString();
            profileBanner = in.readString();
        }

        public String getId() {
            return id;
        }

        public String getDisplayName() {
            return displayName;
        }

        public String getName() {
            return name;
        }

        public String getLogo() {
            return logo;
        }

        public String getProfileBanner() {
            return profileBanner;
        }

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeString(getId());
            dest.writeString(getDisplayName());
            dest.writeString(getName());
            dest.writeString(getLogo());
            dest.writeString(getProfileBanner());
        }

        public static final Creator<Channel> CREATOR = new Creator<Channel>() {
            public Channel createFromParcel(Parcel in) {
                return new Channel(in);
            }

            public Channel[] newArray(int size) {
                return new Channel[size];
            }
        };

        public String toString() {
            return String.format(Locale.getDefault(),
                    "Channel { displayName: %s, name: %s }",
                    getDisplayName(), getName());
        }
    }

    public static class Wrapper implements Parcelable {
        public Stream stream;

        public Wrapper(Parcel in) {
            stream = in.readParcelable(Wrapper.class.getClassLoader());
        }

        public Stream getStream() {
            return stream;
        }

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeParcelable(getStream(), flags);
        }

        public static final Creator<Wrapper> CREATOR = new Creator<Wrapper>() {
            public Wrapper createFromParcel(Parcel in) {
                return new Wrapper(in);
            }

            public Wrapper[] newArray(int size) {
                return new Wrapper[size];
            }
        };
    }

}

File: app/src/main/java/dk/lndesign/relay/model/Stream.kt
package dk.lndesign.relay.model

import android.os.Parcel
import android.os.Parcelable

import com.google.gson.annotations.SerializedName

import java.util.Locale

class Stream(@SerializedName("_id") val id: String,
             val game: String,
             val viewers: Int?,
             val preview: Preview,
             val channel: Channel,
             @SerializedName("stream_type") val streamType: String): Parcelable {

    private constructor(parcel: Parcel): this(
        id = parcel.readString(),
        game = parcel.readString(),
        viewers = parcel.readInt(),
        preview = parcel.readParcelable(Preview::class.java.classLoader),
        channel = parcel.readParcelable(Channel::class.java.classLoader),
        streamType = parcel.readString()
    )

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(id)
        dest.writeString(game)
        dest.writeInt(viewers!!)
        dest.writeParcelable(preview, flags)
        dest.writeParcelable(channel, flags)
        dest.writeString(streamType)
    }

    override fun toString(): String {
        return "Stream { id: $id, game: '$game', viewers: $viewers, channel: $channel }"
    }

    class Preview(val small: String,
                  val medium: String,
                  val large: String) : Parcelable {

        private constructor(parcel: Parcel): this(
                small = parcel.readString(),
                medium = parcel.readString(),
                large = parcel.readString()
        )

        override fun describeContents() = 0

        override fun writeToParcel(dest: Parcel, flags: Int) {
            dest.writeString(small)
            dest.writeString(medium)
            dest.writeString(large)
        }

        override fun toString(): String {
            return "Preview { small: $small, medium: $medium, large: $large }"
        }

        companion object {
            @JvmField val CREATOR = object : Parcelable.Creator<Preview> {
                override fun createFromParcel(parcel: Parcel): Preview {
                    return Preview(parcel)
                }

                override fun newArray(size: Int): Array<Preview?> {
                    return arrayOfNulls(size)
                }
            }
        }
    }

    class Channel(@SerializedName("_id")
                  val id: String,
                  @SerializedName("display_name") val displayName: String,
                  val name: String,
                  val logo: String,
                  @SerializedName("profile_banner") val profileBanner: String) : Parcelable {

        private constructor(parcel: Parcel): this(
                id = parcel.readString(),
                displayName = parcel.readString(),
                name = parcel.readString(),
                logo = parcel.readString(),
                profileBanner = parcel.readString()
        )

        override fun describeContents() = 0

        override fun writeToParcel(dest: Parcel, flags: Int) {
            dest.writeString(id)
            dest.writeString(displayName)
            dest.writeString(name)
            dest.writeString(logo)
            dest.writeString(profileBanner)
        }

        override fun toString(): String {
            return String.format(Locale.getDefault(),
                    "Channel { displayName: %s, name: %s }",
                    displayName, name)
        }

        companion object {
            @JvmField val CREATOR = object : Parcelable.Creator<Channel> {
                override fun createFromParcel(parcel: Parcel): Channel {
                    return Channel(parcel)
                }

                override fun newArray(size: Int): Array<Channel?> {
                    return arrayOfNulls(size)
                }
            }
        }
    }

    class Wrapper(val stream: Stream) : Parcelable {

        private constructor(parcel: Parcel): this(
            stream = parcel.readParcelable(Wrapper::class.java.classLoader)
        )

        override fun describeContents() = 0

        override fun writeToParcel(dest: Parcel, flags: Int) {
            dest.writeParcelable(stream, flags)
        }

        companion object {
            @JvmField val CREATOR = object : Parcelable.Creator<Wrapper> {
                override fun createFromParcel(parcel: Parcel): Wrapper {
                    return Wrapper(parcel)
                }

                override fun newArray(size: Int): Array<Wrapper?> {
                    return arrayOfNulls(size)
                }
            }
        }
    }

    companion object {
        @JvmField val CREATOR = object : Parcelable.Creator<Stream> {
            override fun createFromParcel(parcel: Parcel): Stream {
                return Stream(parcel)
            }

            override fun newArray(size: Int): Array<Stream?> {
                return arrayOfNulls(size)
            }
        }
    }

}
